Avoid too large tracer gradient "slope"

jm-c opened the pull request:
Avoid machine over-flow in computing tracer distribution slope by
putting a cap (here: CrMax =1.e+6) before entering Limiter Fct.
This is similar to what has been done (for some time) in gad_dst3fl_adv_*.F routines.

Note: the precise value of the CrMax does not matter much for any of the
current (in GAD_FLUX_LIMITER.h) Limiter functions.

## What changes does this PR introduce?
supposed to fix issue #459 

## What is the current behaviour? 
Division by "Rj" can result in overflow since it could be very small but not zero.

## What is the new behaviour 
Avoid this division when Rj is too small (compared to neighboring gradient).

## Does this PR introduce a breaking change? 
no

## Other information:


## Suggested addition to `tag-index`
o pkg/generic_advdiff:
  - improve "slope" calculation in DST2 Flux-Limiter advection scheme to avoid
    potential over-flow arithmetic in division by small slope (issue #459),
    similar to what has been done in gad_dst3fl_*.F routines.

menemenlis reviewed (commented):
Much better than what I suggested, I agree.  What about gad_fluxlimit_impl_r.F ?
Can't that also run in same problem?

How do I get rid of PR #461 ?

jm-c commented:
@menemenlis  You don't need to get rid of PR #461 right  now. This PR is not finished (tag-index addition missing, gad_fluxlimit_impl_r.F to fix and I have not tested the adjoint yet but I expect it will pass).

menemenlis reviewed (approved):
Very nice!

A stylistic comment: sometimes these routines use zeroRL, sometimes 0., and sometimes 0. _d 0.  Similarly for oneRL and 1. _d 0.  Is there a reason for not using exclusively zeroRL and oneRL?

jm-c commented:
> Very nice!
> 
> A stylistic comment: sometimes these routines use zeroRL, sometimes 0., and sometimes 0. _d 0. Similarly for oneRL and 1. _d 0. Is there a reason for not using exclusively zeroRL and oneRL?

Yes, it's not super clear:
1) it's better to have "0. _d 0" rather than just "0." (in expression and setting) since the later will have to be converted to real*8; 
  note that I never tried to test if these conversions were slowing down the calculation.
  But with integer and negative power of 2 the numerical result will be the same after conversion.
  Now when it's not an integer (e.g., _RL varRL, varRL = 2.1 , instead of varRL = 2.1 _d 0 ) the numerical result will be different
  and the difference will be compiler dependent.
2) I added "zeroRS" plus few other because, when using an intrinsic function with more than 1 argument, 
   the compiler does not know which version to pick when the 2 (or more) arguments are not the same type. 
  And if varRS is _RS,
      varRS = MAX( varRS, 0. _d 0 ) will cause problem when _RS is real*4
   and:
      varRS = MAX( varRS, 0. )      will cause problem when _RS is real*8
   but:
      varRS = MAX( varRS, zeroRS )  will always work.
   And just for convienience (+ avoid thinking too much when coding), I also  added "zeroRL" and few others 
   but using these later ones is the same as using 0. _d 0 (or 1. _d 0 for oneRL, ...).

mjlosch reviewed (approved):
This looks good to me. I see the analogy to other schemes (`gad_dst3fl_*`), AD-code looks OK, not extra complaints by taf (although there are some for `gad_dst3fl_impl_r` and I would have expected something similar here for `gad_fluxlimit_impl_r`, but I guess it's less complicated).

jm-c commented:
Going to merge this PR later today.
